### PR TITLE
UCP/PROTO: Move non-rndv protocols to use probe() instead of init()

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Next phase of adding protocol variants, moving from init() to probe()

### CODE REVIEW

**Comment by yosefe** [comment]
@ivankochin @brminich can you pls review?

File: src/ucp/proto/proto_multi.c [hunk 1]
```diff
@@ -242,17 +237,44 @@ ucs_status_t ucp_proto_multi_init(const ucp_proto_multi_init_params_t *params,
     return status;
 }
 
+size_t
+ucp_proto_multi_priv_size(const ucp_proto_multi_priv_t *mpriv, size_t full_size)
+{
+    ucs_assert(full_size >= sizeof(ucp_proto_multi_priv_t));
```
Comment (brminich) [inline src/ucp/proto/proto_multi.c:243]: looks hacky, is it possible to calculate just `mpriv->num_lanes*ucs_field_sizeof(ucp_proto_multi_priv_t, lanes[0])` and the rest calculate separatly in each place?

**Comment by ivankochin** [comment]
Not directly connected to this patch, but anyway. As I understand, at the current moment, failing of probing or initialization doesn't  entail any traces printing. Maybe it worth adding some logging which would print status of  initialization or how many variants were added by probe? e.g. extend `ucp_proto_select_add_proto`

File: src/ucp/proto/proto_multi.h [hunk 2]
```diff
@@ -18,6 +18,33 @@
...
+    ({ \
+        typedef ucs_typeof(*(_priv)) _type; \
+        \
+        /* Make sure _mpriv is the last field in _type */ \
+        UCS_STATIC_ASSERT((ucs_offsetof(_type, _mpriv) + \
+                           sizeof(ucp_proto_multi_priv_t)) == sizeof(_type)); \
+        \
+        /* Subtract the full size of _mpriv and add the actual size. */ \
+        sizeof(_type) - sizeof(ucp_proto_multi_priv_t) + \
+                ucp_proto_multi_priv_size(&(_priv)->_mpriv); \
```
Comment (brminich) [inline src/ucp/proto/proto_multi.h:44]: maybe `ucs_offsetof(_type, _mpriv) + ucp_proto_multi_priv_size(&(_priv)->_mpriv)`?

File: src/ucp/proto/proto_multi.h [hunk 3]
```diff
@@ -65,7 +95,7 @@ typedef struct {
     ucp_lane_index_t            num_lanes;    /* Number of lanes to use */
     size_t                      align_thresh; /* Cached value of threshold for
                                                  enabling data split alignment */
-    ucp_proto_multi_lane_priv_t lanes[0];     /* Array of lanes */
+    ucp_proto_multi_lane_priv_t lanes[UCP_MAX_LANES]; /* Array of lanes */
```
Comment (ivankochin) [inline src/ucp/proto/proto_multi.h:98]: Why changed to `UCP_MAX_LENGTH`? 

Minor: Align comments for other fields.

File: src/ucp/proto/proto_select.c [hunk 4]
```diff
@@ -716,8 +716,10 @@ void ucp_proto_select_add_proto(const ucp_proto_init_params_t *init_params,
     ucp_proto_id_t proto_id                       = init_params->proto_id;
     ucp_proto_init_elem_t *init_elem;
     const char *proto_name;
-    size_t priv_offset;
+    char min_length_str[64];
+    char thresh_str[64];
```
Comment (ivankochin) [inline src/ucp/proto/proto_select.c:721]: Extra blank line?

File: src/ucp/proto/proto_multi.h (hunk 3 above)
Comment (ivankochin) [reply src/ucp/proto/proto_multi.h:98]: I thought that rpriv is always `UCP_PROTO_PRIV_MAX` buffer allocated on the `ucp_proto_select _init_protocols` side. Maybe I am missing something, but I cannot find the place where we allocate `ucp_proto_multi_priv_t` using alloca.

Is that change made for some future improvements?

Comment (yosefe) [reply src/ucp/proto/proto_multi.h:98]: For example here: https://github.com/openucx/ucx/pull/9695/files#diff-a430ba778d55dee29e9281cc476b4563ea1732a2462db82afb8dcd1c0bd1972aR249 we place ucp_proto_multi_priv_t  on the stack; if it had zero-size array of lanes we could not do that - we had to allocate with alloca and add space for the lanes (which create more code).

**Review by brminich — DISMISSED** [review]

**Review by ivankochin — APPROVED** [review]

File: src/ucp/proto/proto_multi.h (hunk 3 above)
Comment (ivankochin) [reply src/ucp/proto/proto_multi.h:98]: Understood, thanks!

File: src/ucp/proto/proto_select.c (hunk 4 above)
Comment (yosefe) [reply src/ucp/proto/proto_select.c]: > Extra blank line?

fixed

**Pull request merged**